### PR TITLE
Pat 928 prisoner no search endpoint

### PR DESCRIPTION
**Pull request by lbennett-moj**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -22,14 +22,14 @@ Any commands in `localstack/setup-sns.sh` will be run when `localstack` starts,
...
-TMPDIR=/private$TMPDIR docker-compose up 
+TMPDIR=/private$TMPDIR docker-compose up
 ```
 Queues and topics will automatically be created when the `localstack` container starts.
 
 Running all services except this application (hence allowing you to run this in the IDE)
 
 ```bash
-TMPDIR=/private$TMPDIR docker-compose up --scale prisoner-offender-search=0 
+TMPDIR=/private$TMPDIR docker-compose up --scale prisoner-offender-search=0
```
Comment (lbennett-moj) [inline README.md:32]: Ignore these whitespace changes -  I changed my mind about an edit and for some reason I can't replace it with the original version

File: README.md [hunk 2]
```diff
@@ -77,7 +77,7 @@ curl --location --request POST "http://localhost:4571/prisoner-search-a/_search"
 Or to just run `localstack` which is useful when running against an a non-local test system Env need to be `spring.profiles.active=localstack` and `sqs.provider=full-localstack`
 
 ```bash
-TMPDIR=/private$TMPDIR docker-compose up localstack 
+TMPDIR=/private$TMPDIR docker-compose up localstack
 ```
```
Comment (lbennett-moj) [inline README.md:81]: sorry - can't get rid of these annoying whitespace changes for some reason

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/PrisonerSearchService.kt [hunk 3]
```diff
@@ -57,6 +57,24 @@ class PrisonerSearchService(
     } ?: Result.NoMatch
   }
 
+  private fun queryBy(searchCriteria: PrisonerListCriteria, queryBuilder: (searchCriteria: PrisonerListCriteria) -> BoolQueryBuilder?): Result {
+    val query = queryBuilder(searchCriteria)
+    return query?.let {
+      val searchSourceBuilder = SearchSourceBuilder().apply {
+        query(query)
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/PrisonerSearchService.kt:64]: you could use `query(it)`  - but both are readable so probably doesn't matter

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/PrisonerSearchService.kt [hunk 4]
```diff
@@ -57,6 +57,24 @@ class PrisonerSearchService(
...
+      }
+      val searchRequest = SearchRequest(arrayOf(getIndex()), searchSourceBuilder)
+      val prisonerMatches = getSearchResult(searchClient.search(searchRequest))
+      return if (prisonerMatches.isEmpty()) Result.NoMatch else Result.Match(prisonerMatches)
+    } ?: Result.NoMatch
+  }
+
+  private fun matchByIds(searchCriteria: PrisonerListCriteria): BoolQueryBuilder? {
+    with(searchCriteria) {
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/PrisonerSearchService.kt:73]: looks like the `with` is redundant here, you tend to use in Kotlin so you don't have to `searchCriteria.prisonerNumbers` and use `prisonerNumbers`, given there is one statement, I would remove the `with`

File: src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/resource/PrisonerSearchResourceTest.kt [hunk 5]
```diff
@@ -60,6 +61,80 @@ class PrisonerSearchResourceTest : QueueIntegrationTest() {
...
+      .headers(setAuthorisation(roles = listOf("ROLE_GLOBAL_SEARCH")))
+      .header("Content-Type", "application/json")
+      .exchange()
+      .expectStatus().isOk
+      .expectBody().json("/results/search_results_A7089FA.json".readResourceAsText())
+  }
+
+  private fun getTestprisonerNumbers(count: Int): List<String> {
+    return List(count) { i -> "AN$i"  }
+  }
```
Comment (andymarke) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/resource/PrisonerSearchResourceTest.kt:136]: Nice, not seen that syntax before.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/PrisonerSearchService.kt [hunk 6]
```diff
@@ -105,6 +123,16 @@ class PrisonerSearchService(
   private fun getIndex(): String{
     return indexStatusService.getCurrentIndex().currentIndex.indexName
   }
+
+  fun findByListOfprisonerNumbers(prisonerListCriteria: PrisonerListCriteria): List<Prisoner> {
```
Comment (yasinmustafa) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/PrisonerSearchService.kt:127]: Slight typo in the function name - should be camel case; the 'P' in prisoner should be uppercase

File: src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/resource/PrisonerSearchResourceTest.kt [hunk 7]
```diff
@@ -60,6 +61,80 @@ class PrisonerSearchResourceTest : QueueIntegrationTest() {
...
+    webTestClient.post().uri("/prisoner-search/prisoner-numbers")
+      .body(BodyInserters.fromValue(gson.toJson(PrisonerListCriteria(listOf("A7089FA", "A709999", "A707777")))))
+      .headers(setAuthorisation(roles = listOf("ROLE_GLOBAL_SEARCH")))
+      .header("Content-Type", "application/json")
+      .exchange()
+      .expectStatus().isOk
+      .expectBody().json("/results/search_results_A7089FA.json".readResourceAsText())
+  }
+
+  private fun getTestprisonerNumbers(count: Int): List<String> {
```
Comment (yasinmustafa) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/resource/PrisonerSearchResourceTest.kt:134]: Funcation name typo again, priosoner should be Prisoner

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/PrisonerListCriteria.kt [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties
+import io.swagger.annotations.ApiModel
+import io.swagger.annotations.ApiModelProperty
+
+@ApiModel("Search Criteria for a list of prisoners")
+
+@JsonIgnoreProperties( value =  [ "isValid" ])
+data class PrisonerListCriteria (
+  @ApiModelProperty(required = false, value = "List of prisoner numbers to search by", example = "A1234AA,", position = 1)
```
Comment (yasinmustafa) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/PrisonerListCriteria.kt:11]: minor but shouldn't the example here be an array; e.g. [ 'A1234A' ]

**Pull request merged**